server/src/main/java/org/elasticsearch/index/query/InterceptedQueryBuilderWrapper.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+/**
+ * Wrapper for instances of {@link AbstractQueryBuilder} that have been intercepted using the {@link QueryRewriteInterceptor} to
+ * break out of the rewrite phase. These instances are unwrapped on serialization.
+ * @param <T>
+ */
+public class InterceptedQueryBuilderWrapper<T extends AbstractQueryBuilder<T>> extends AbstractQueryBuilder<T> {
```

Comment:
Thanks, I'll look into adding tests after I complete the refactoring!

Reply:
Why do we need a template here? Why not a simple `InterceptedQueryBuilderWrapper extends AbstractQueryBuilder<InterceptedQueryBuilderWrapper>`?

Reply:
I also think it would be easier to extends a `QueryBuilder` rather than the `AbstractQueryBuilder` to do the wrapping so that we can call `toQuery` and not the `doTo...` versions.

Reply:
Refactored to `implements QueryBuilder`